File: backend-kotlin/src/main/kotlin/com/example/modules/catalog/internal/CatalogService.kt
// Business logic — hidden from the outside world.
package com.example.modules.catalog.internal

import com.example.modules.catalog.infrastructure.Product
import com.example.modules.catalog.infrastructure.seedProducts

class CatalogService {
    private val products = seedProducts()

    fun list(category: String?) = products.filter { category.isNullOrEmpty() || it.category == category }
    fun get(sku: String): Product? = products.find { it.sku == sku }
    fun search(query: String): List<Product> {
        val q = query.lowercase()
        return products.filter { it.name.lowercase().contains(q) || it.description.lowercase().contains(q) }
    }
}

File: backend-scala/src/main/scala/com/example/modules/catalog/internal/CatalogService.scala
// Business logic — hidden from the outside world.
package com.example.modules.catalog.internal

import com.example.modules.catalog.infrastructure.{Product, SeedData}

class CatalogService:
  private val products = SeedData.seed

  def list(category: Option[String]): List[Product] =
    products.filter(p => category.forall(c => c.isEmpty || p.category == c))

  def get(sku: String): Option[Product] = products.find(_.sku == sku)

  def search(query: String): List[Product] =
    val q = query.toLowerCase
    products.filter(p => p.name.toLowerCase.contains(q) || p.description.toLowerCase.contains(q))
